File: src/session/session-summary.ts
import { rawTextFromContent } from "./content-text";
import type {
  TimelineEntry,
  ToolCall,
  ToolCallContentItem,
  ToolKind,
} from "./types";

/** ACP kinds that typically mutate the workspace. */
const MUTATING_KINDS = new Set<ToolKind>(["edit", "delete", "move"]);

/**
 * Titles used by Claude Code / common agents when kind is missing or "other".
 * ACP often labels Write/Edit as kind "other" without a structured diff.
 */
const WRITE_TITLE_RE =
  /^(Write|Edit|Create|Update|Delete|Remove|Move|Rename|Apply[_\s-]?Patch|StrReplace|MultiEdit|NotebookEdit)\b/i;

/** Shell commands that usually mutate files (when kind is execute). */
const EXEC_MUTATE_RE =
  /\b(rm|rmdir|mv|cp|install|tee|truncate|chmod|chown|touch|mkdir|ln|sed\s+-i|perl\s+-i|git\s+(apply|checkout|restore|rm|mv|add|commit|reset|rebase|merge|cherry-pick|stash)|npm\s+i(?:nstall)?|pnpm\s+i(?:nstall)?|yarn\s+add|bun\s+add|pip\s+install|cargo\s+add)\b/i;

export type FileChangeKind = "edit" | "delete" | "move" | "other";

export type FileChange = {
  path: string;
  kind: FileChangeKind;
  /** Truncated unified-ish diff text when available. */
  diff?: string;
  /** Extra notes (e.g. shell command that mutated the path). */
  note?: string;
  toolTitle?: string;
  status?: string;
};

export type SessionChangeSummary = {
  /** Full markdown summary for seeding a review session. */
  text: string;
  /**
   * True when the session has file mutations (edit/delete/move, diffs,
   * write-like tools, or mutating shell commands).
   */
  hasChanges: boolean;
  /**
   * True when there is anything useful to review in a new session
   * (user goals, tools, or agent replies) — not only structured file edits.
   */
  hasReviewableContent: boolean;
  fileCount: number;
  files: FileChange[];
  userGoals: string[];
};

export type SummarizeSessionOptions = {
  sessionTitle?: string;
  project?: string;
  /** Max lines included per file diff (default 80). */
  maxDiffLines?: number;
  /** Max total characters for the markdown body (default 14000). */
  maxChars?: number;
  /** Max user goals to include (default 8). */
  maxGoals?: number;
  /** Max agent note snippets (default 3). */
  maxAgentNotes?: number;
};

const DEFAULT_MAX_DIFF_LINES = 80;
const DEFAULT_MAX_CHARS = 14_000;
const DEFAULT_MAX_GOALS = 8;
const DEFAULT_MAX_AGENT_NOTES = 3;

/**
 * Default first user prompt for a review session. This is the primary
 * requirement the new-session agent should satisfy.
 */
export const DEFAULT_REVIEW_PROMPT = [
  "Review the work from the previous session (summarized in the starting context).",
  "",
  "Focus on:",
  "1. Correctness — logic bugs, edge cases, broken contracts",
  "2. Security — injection, secrets, unsafe paths or permissions",
  "3. Regressions — unintended side effects on existing behavior",
  "4. Tests — missing or weak coverage for the changes",
  "5. Code quality — clarity, maintainability, unnecessary complexity",
  "",
  "For each finding report: severity (blocker / major / minor / nit), file location, what's wrong, and a concrete fix suggestion. End with an overall assessment.",
  "If the summary has no structured file diffs, inspect the project (and git status/diff if available) against the user goals and tool activity listed above.",
].join("\n");

function truncateLines(text: string, maxLines: number): string {
  const lines = text.split("\n");
  if (lines.length <= maxLines) return text;
  const kept = lines.slice(0, maxLines);
  const omitted = lines.length - maxLines;
  return `${kept.join("\n")}\n… (${omitted} more lines truncated)`;
}

function truncateChars(text: string, maxChars: number): string {
  if (text.length <= maxChars) return text;
  return `${text.slice(0, Math.max(0, maxChars - 40)).trimEnd()}\n\n… (summary truncated)`;
}

function asRecord(value: unknown): Record<string, unknown> | null {
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }
  return null;
}

function stringField(
  obj: Record<string, unknown> | null,
  ...keys: string[]
): string | null {
  if (!obj) return null;
  for (const k of keys) {
    const v = obj[k];
    if (typeof v === "string" && v.trim()) return v;
  }
  return null;
}

function pathFromRawInput(tc: ToolCall): string | null {
  const input = asRecord(tc.rawInput);
  return stringField(
    input,
    "path",
    "file_path",
    "filePath",
    "target",
    "dest",
    "to",
  );
}

function pathFromToolCall(tc: ToolCall): string | null {
  for (const item of tc.content) {
    if (item.type === "diff" && item.path) return item.path;
  }
  const loc = tc.locations?.[0]?.path;
  if (loc) return loc;
  const fromInput = pathFromRawInput(tc);
  if (fromInput) return fromInput;

  // Heuristic: titles like "Edit foo.ts", "Write src/a.ts", "rm ./ui.html"
  const m = tc.title.match(
    /(?:^|\s)((?:\/|\.\/)?[\w.@+-]+(?:\/[\w.@+-]+)*(?:\.\w+)?)\s*$/,
  );
  return m?.[1] ?? null;
}

function contentSnippetFromRawInput(
  tc: ToolCall,
  maxDiffLines: number,
): string | undefined {
  const input = asRecord(tc.rawInput);
  if (!input) return undefined;
  const content = stringField(
    input,
    "content",
    "contents",
    "new_string",
    "newString",
    "new_text",
    "newText",
  );
  const old = stringField(input, "old_string", "oldString", "old_text", "oldText");
  if (old && content) {
    return [
      "--- before",
      truncateLines(old, Math.floor(maxDiffLines / 2)),
      "+++ after",
      truncateLines(content, Math.floor(maxDiffLines / 2)),
    ].join("\n");
  }
  if (content) return truncateLines(content, maxDiffLines);
  return undefined;
}

function kindFromToolCall(tc: ToolCall): FileChangeKind {
  if (tc.kind === "edit" || tc.kind === "delete" || tc.kind === "move") {
    return tc.kind;
  }
  if (tc.content.some((c) => c.type === "diff")) return "edit";
  const title = tc.title;
  if (/^(Delete|Remove|rm\b)/i.test(title)) return "delete";
  if (/^(Move|Rename|mv\b)/i.test(title)) return "move";
  if (WRITE_TITLE_RE.test(title)) return "edit";
  if (tc.kind === "execute" && EXEC_MUTATE_RE.test(title)) {
    if (/\brm\b|\brmdir\b|git\s+rm\b/i.test(title)) return "delete";
    if (/\bmv\b|git\s+mv\b/i.test(title)) return "move";
    return "edit";
  }
  const input = asRecord(tc.rawInput);
  if (input && stringField(input, "content", "contents", "new_string", "newString")) {
    return "edit";
  }
  return "other";
}

function formatDiffItem(
  item: Extract<ToolCallContentItem, { type: "diff" }>,
  maxDiffLines: number,
): string {
  const oldText = item.oldText ?? "";
  const newText = item.newText ?? "";
  let body: string;
  if (oldText && newText) {
    body = [
      "--- before",
      truncateLines(oldText, Math.floor(maxDiffLines / 2)),
      "+++ after",
      truncateLines(newText, Math.floor(maxDiffLines / 2)),
    ].join("\n");
  } else if (newText) {
    body = truncateLines(newText, maxDiffLines);
  } else if (oldText) {
    body = truncateLines(oldText, maxDiffLines);
  } else {
    body = "(empty diff)";
  }
  return body;
}

/** True when this tool call likely mutated files (including Claude ACP quirks). */
export function isMutatingToolCall(tc: ToolCall): boolean {
  if (tc.kind && MUTATING_KINDS.has(tc.kind)) return true;
  if (tc.content.some((c) => c.type === "diff")) return true;
  if (WRITE_TITLE_RE.test(tc.title)) return true;

  const input = asRecord(tc.rawInput);
  if (input) {
    const hasPath = Boolean(
      stringField(input, "path", "file_path", "filePath", "target"),
    );
    const hasWriteBody = Boolean(
      stringField(
        input,
        "content",
        "contents",
        "new_string",
        "newString",
        "old_string",
        "oldString",
      ),
    );
    // Write/Edit-style tools often send file_path + content without kind=edit.
    if (hasPath && hasWriteBody) return true;
  }

  if (tc.kind === "execute" && EXEC_MUTATE_RE.test(tc.title)) return true;
  // Some adapters leave execute commands as kind "other".
  if ((!tc.kind || tc.kind === "other") && EXEC_MUTATE_RE.test(tc.title)) {
    return true;
  }
  return false;
}

function upsertFileChange(
  fileMap: Map<string, FileChange>,
  next: FileChange,
): void {
  const existing = fileMap.get(next.path);
  if (!existing) {
    fileMap.set(next.path, next);
    return;
  }
  fileMap.set(next.path, {
    path: next.path,
    kind: existing.kind === "delete" || next.kind === "delete" ? "delete" : next.kind,
    toolTitle: next.toolTitle ?? existing.toolTitle,
    status: next.status ?? existing.status,
    note: [existing.note, next.note].filter(Boolean).join("\n") || undefined,
    diff: next.diff
      ? existing.diff
        ? `${existing.diff}\n\n${next.diff}`
        : next.diff
      : existing.diff,
  });
}

/**
 * Build a structured, markdown change summary from a session timeline.
 * Pure — safe to call from the webview without Bun.
 */
export function summarizeSessionChanges(
  timeline: TimelineEntry[],
  opts: SummarizeSessionOptions = {},
): SessionChangeSummary {
  const maxDiffLines = opts.maxDiffLines ?? DEFAULT_MAX_DIFF_LINES;
  const maxChars = opts.maxChars ?? DEFAULT_MAX_CHARS;
  const maxGoals = opts.maxGoals ?? DEFAULT_MAX_GOALS;
  const maxAgentNotes = opts.maxAgentNotes ?? DEFAULT_MAX_AGENT_NOTES;

  const userGoals: string[] = [];
  const agentNotes: string[] = [];
  const fileMap = new Map<string, FileChange>();
  const otherTools: string[] = [];
  let toolCount = 0;

  for (const entry of timeline) {
    if (entry.type === "message" && entry.role === "user") {
      const text = rawTextFromContent(entry.content).trim();
      if (text && userGoals.length < maxGoals) {
        const oneLine = text.replace(/\s+/g, " ");
        userGoals.push(
          oneLine.length > 240 ? `${oneLine.slice(0, 239)}…` : oneLine,
        );
      }
      continue;
    }

    if (entry.type === "message" && entry.role === "agent") {
      const text = rawTextFromContent(entry.content).trim();
      if (text && agentNotes.length < maxAgentNotes) {
        const oneLine = text.replace(/\s+/g, " ");
        agentNotes.push(
          oneLine.length > 320 ? `${oneLine.slice(0, 319)}…` : oneLine,
        );
      }
      continue;
    }

    if (entry.type !== "tool_call") continue;
    toolCount += 1;
    const tc = entry.toolCall;

    if (isMutatingToolCall(tc)) {
      const path = pathFromToolCall(tc) ?? `(tool) ${tc.title.slice(0, 80)}`;
      const kind = kindFromToolCall(tc);
      const diffs = tc.content.filter(
        (c): c is Extract<ToolCallContentItem, { type: "diff" }> =>
          c.type === "diff",
      );
      let diff: string | undefined;
      if (diffs.length > 0) {
        diff = truncateLines(
          diffs.map((d) => formatDiffItem(d, maxDiffLines)).join("\n\n"),
          maxDiffLines * 2,
        );
      } else {
        diff = contentSnippetFromRawInput(tc, maxDiffLines);
      }
      const note =
        tc.kind === "execute" || EXEC_MUTATE_RE.test(tc.title)
          ? `Shell: ${tc.title}`
          : undefined;
      upsertFileChange(fileMap, {
        path,
        kind,
        toolTitle: tc.title,
        status: tc.status,
        diff,
        note,
      });
    } else {
      // Reads, searches, MCP tools, non-mutating executes — keep a short log.
      if (
        tc.kind === "execute" ||
        tc.kind === "other" ||
        tc.kind === "search" ||
        tc.kind === "fetch" ||
        !tc.kind
      ) {
        const status = tc.status === "failed" ? " (failed)" : "";
        otherTools.push(`- ${tc.title}${status}`);
      }
    }
  }

  const files = [...fileMap.values()];
  const hasChanges = files.length > 0;
  const hasReviewableContent =
    hasChanges ||
    userGoals.length > 0 ||
    toolCount > 0 ||
    agentNotes.length > 0;

  const sections: string[] = [];
  sections.push("# Session work summary");
  if (opts.sessionTitle) {
    sections.push(`**Session:** ${opts.sessionTitle}`);
  }
  if (opts.project) {
    sections.push(`**Project:** ${opts.project}`);
  }
  sections.push("");

  if (userGoals.length > 0) {
    sections.push("## User goals");
    for (const g of userGoals) sections.push(`- ${g}`);
    sections.push("");
  }

  if (files.length > 0) {
    sections.push(`## Files changed (${files.length})`);
    for (const f of files) {
      sections.push(`### \`${f.path}\` (${f.kind})`);
      if (f.toolTitle) sections.push(`_Tool:_ ${f.toolTitle}`);
      if (f.note) sections.push(`_${f.note}_`);
      if (f.diff) {
        sections.push("```");
        sections.push(truncateLines(f.diff, maxDiffLines * 2));
        sections.push("```");
      } else {
        sections.push(
          "_No diff content captured in the session (inspect the file / git diff)._",
        );
      }
      sections.push("");
    }
  } else {
    sections.push("## Files changed");
    sections.push(
      "_No structured file edits were recorded. If the agent used shell commands or tools without diffs, use git status/diff and the activity below._",
    );
    sections.push("");
  }

  if (otherTools.length > 0) {
    sections.push("## Other tool activity");
    const shown = otherTools.slice(0, 30);
    sections.push(...shown);
    if (otherTools.length > 30) {
      sections.push(`- … and ${otherTools.length - 30} more`);
    }
    sections.push("");
  }

  if (agentNotes.length > 0) {
    sections.push("## Agent notes (excerpts)");
    for (const n of agentNotes) sections.push(`- ${n}`);
    sections.push("");
  }

  const text = truncateChars(sections.join("\n").trim(), maxChars);

  return {
    text,
    hasChanges,
    hasReviewableContent,
    fileCount: files.length,
    files,
    userGoals,
  };
}

/** Short title for a review session forked from a source chat. */
export function reviewSessionTitle(sourceTitle?: string, max = 56): string {
  const base = (sourceTitle || "session").replace(/\s+/g, " ").trim();
  const prefix = "Review: ";
  const budget = max - prefix.length;
  if (base.length <= budget) return `${prefix}${base}`;
  return `${prefix}${base.slice(0, Math.max(1, budget - 1))}…`;
}

File: crates/engine/src/session_summary.rs
use std::collections::BTreeMap;

use comet_doc::{MessagePart, MessageRole, SessionMessageEntry};
use comet_proto::ToolCall;

const DEFAULT_MAX_CHARS: usize = 14_000;
const DEFAULT_MAX_GOALS: usize = 8;
const DEFAULT_MAX_AGENT_NOTES: usize = 3;

pub const DEFAULT_REVIEW_PROMPT: &str = "Review the work from the previous session (summarized in the starting context).\n\nFocus on:\n1. Correctness — logic bugs, edge cases, broken contracts\n2. Security — injection, secrets, unsafe paths or permissions\n3. Regressions — unintended side effects on existing behavior\n4. Tests — missing or weak coverage for the changes\n5. Code quality — clarity, maintainability, unnecessary complexity\n\nFor each finding report: severity (blocker / major / minor / nit), file location, what's wrong, and a concrete fix suggestion. End with an overall assessment.\nIf the summary has no structured file diffs, inspect the project (and git status/diff if available) against the user goals and tool activity listed above.";

pub struct SessionChangeSummary {
    pub text: String,
    pub has_changes: bool,
    pub has_reviewable_content: bool,
}

fn message_text(entry: &SessionMessageEntry) -> String {
    entry
        .parts
        .iter()
        .filter_map(|part| match part {
            MessagePart::Text { text, .. } => Some(text.as_str()),
            _ => None,
        })
        .collect::<Vec<_>>()
        .join("\n\n")
}

fn compact(text: &str, max: usize) -> String {
    let one_line = text.split_whitespace().collect::<Vec<_>>().join(" ");
    if one_line.chars().count() <= max {
        one_line
    } else {
        format!(
            "{}…",
            one_line
                .chars()
                .take(max.saturating_sub(1))
                .collect::<String>()
        )
    }
}

fn exec_mutates(command: &str) -> bool {
    let padded = format!(" {} ", command.to_ascii_lowercase());
    [
        " rm ",
        " rmdir ",
        " mv ",
        " cp ",
        " install ",
        " tee ",
        " truncate ",
        " chmod ",
        " chown ",
        " touch ",
        " mkdir ",
        " ln ",
        " sed -i ",
        " git apply ",
        " git checkout ",
        " git restore ",
        " git rm ",
        " git mv ",
        " git add ",
        " git commit ",
        " git reset ",
        " npm install ",
        " npm i ",
        " pnpm install ",
        " yarn add ",
        " bun add ",
        " pip install ",
        " cargo add ",
    ]
    .iter()
    .any(|needle| padded.contains(needle))
}

fn tool_summary(call: &ToolCall) -> (String, Option<(&str, &str)>) {
    match call {
        ToolCall::Exec { command } => (
            format!("Shell: {}", compact(command, 160)),
            exec_mutates(command).then_some(("(shell command)", "edit")),
        ),
        ToolCall::ReadFile { path } => (format!("Read {path}"), None),
        ToolCall::WriteFile { path, .. } => (format!("Write {path}"), Some((path, "edit"))),
        ToolCall::EditFile { path, .. } => (format!("Edit {path}"), Some((path, "edit"))),
        ToolCall::ApplyPatch { path } => {
            let path = path.as_deref().unwrap_or("(apply patch)");
            (format!("Apply patch {path}"), Some((path, "edit")))
        }
        ToolCall::Search { pattern, path } => (
            format!(
                "Search {}{}",
                pattern,
                path.as_deref()
                    .map(|p| format!(" in {p}"))
                    .unwrap_or_default()
            ),
            None,
        ),
        ToolCall::Glob { pattern } => (format!("Glob {pattern}"), None),
        ToolCall::WebFetch { url, .. } => (format!("Fetch {url}"), None),
        ToolCall::WebSearch { query } => (format!("Web search {query}"), None),
        ToolCall::Todo { .. } => ("Update todo list".into(), None),
        ToolCall::Mcp { server, tool, .. } => (format!("{server}: {tool}"), None),
        ToolCall::Unknown { name, .. } => (name.clone(), None),
    }
}

pub fn summarize_session_changes(
    entries: &[SessionMessageEntry],
    session_title: Option<&str>,
    project: Option<&str>,
) -> SessionChangeSummary {
    let mut user_goals = Vec::new();
    let mut agent_notes = Vec::new();
    let mut files: BTreeMap<String, String> = BTreeMap::new();
    let mut other_tools = Vec::new();
    let mut tool_count = 0usize;

    for entry in entries {
        let text = message_text(entry);
        if entry.role == MessageRole::User
            && !text.trim().is_empty()
            && user_goals.len() < DEFAULT_MAX_GOALS
        {
            user_goals.push(compact(&text, 240));
        } else if entry.role == MessageRole::Assistant
            && !text.trim().is_empty()
            && agent_notes.len() < DEFAULT_MAX_AGENT_NOTES
        {
            agent_notes.push(compact(&text, 320));
        }
        for part in &entry.parts {
            let MessagePart::Tool { call, is_error, .. } = part else {
                continue;
            };
            tool_count += 1;
            let (label, mutation) = tool_summary(call);
            if let Some((path, kind)) = mutation {
                files.insert(path.to_string(), kind.to_string());
            } else {
                other_tools.push(format!(
                    "- {label}{}",
                    if *is_error { " (failed)" } else { "" }
                ));
            }
        }
    }

    let has_changes = !files.is_empty();
    let has_reviewable_content =
        has_changes || !user_goals.is_empty() || tool_count > 0 || !agent_notes.is_empty();
    let mut sections = vec!["# Session work summary".to_string()];
    if let Some(title) = session_title.filter(|title| !title.trim().is_empty()) {
        sections.push(format!("**Session:** {title}"));
    }
    if let Some(project) = project.filter(|project| !project.trim().is_empty()) {
        sections.push(format!("**Project:** {project}"));
    }
    sections.push(String::new());

    if !user_goals.is_empty() {
        sections.push("## User goals".into());
        sections.extend(user_goals.iter().map(|goal| format!("- {goal}")));
        sections.push(String::new());
    }

    sections.push(if files.is_empty() {
        "## Files changed".into()
    } else {
        format!("## Files changed ({})", files.len())
    });
    if files.is_empty() {
        sections.push("_No structured file edits were recorded. If the agent used shell commands or tools without diffs, use git status/diff and the activity below._".into());
    } else {
        for (path, kind) in &files {
            sections.push(format!("### `{path}` ({kind})"));
            sections.push(
                "_No diff content is retained in the transcript; inspect the file and git diff._"
                    .into(),
            );
        }
    }
    sections.push(String::new());

    if !other_tools.is_empty() {
        sections.push("## Other tool activity".into());
        sections.extend(other_tools.iter().take(30).cloned());
        if other_tools.len() > 30 {
            sections.push(format!("- … and {} more", other_tools.len() - 30));
        }
        sections.push(String::new());
    }
    if !agent_notes.is_empty() {
        sections.push("## Agent notes (excerpts)".into());
        sections.extend(agent_notes.iter().map(|note| format!("- {note}")));
    }

    let mut text = sections.join("\n").trim().to_string();
    if text.chars().count() > DEFAULT_MAX_CHARS {
        text = format!(
            "{}\n\n… (summary truncated)",
            text.chars()
                .take(DEFAULT_MAX_CHARS - 40)
                .collect::<String>()
                .trim_end()
        );
    }
    SessionChangeSummary {
        text,
        has_changes,
        has_reviewable_content,
    }
}

pub fn review_session_title(source_title: Option<&str>, max: usize) -> String {
    let base = source_title
        .unwrap_or("session")
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    let prefix = "Review: ";
    let budget = max.saturating_sub(prefix.chars().count());
    if base.chars().count() <= budget {
        format!("{prefix}{base}")
    } else {
        format!(
            "{prefix}{}…",
            base.chars()
                .take(budget.saturating_sub(1).max(1))
                .collect::<String>()
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use comet_doc::MessageStatus;

    fn entry(role: MessageRole, text: &str, parts: Vec<MessagePart>) -> SessionMessageEntry {
        let mut all = vec![MessagePart::Text {
            id: "t0".into(),
            text: text.into(),
        }];
        all.extend(parts);
        SessionMessageEntry {
            id: "m".into(),
            role,
            parts: all,
            created_at: 1,
            device_id: "d".into(),
            status: Some(MessageStatus::Complete),
            continuation_of: None,
        }
    }

    #[test]
    fn summary_collects_goals_changes_and_notes() {
        let entries = vec![
            entry(MessageRole::User, "Fix the parser", vec![]),
            entry(
                MessageRole::Assistant,
                "Implemented and tested it.",
                vec![MessagePart::Tool {
                    id: "tool".into(),
                    call: ToolCall::EditFile {
                        path: "src/parser.rs".into(),
                        old_string: None,
                        new_string: None,
                    },
                    is_error: false,
                    resolved: true,
                }],
            ),
        ];
        let summary = summarize_session_changes(&entries, Some("Parser"), Some("comet"));
        assert!(summary.has_changes && summary.has_reviewable_content);
        assert!(summary.text.contains("Fix the parser"));
        assert!(summary.text.contains("`src/parser.rs`"));
        assert!(summary.text.contains("Implemented and tested it"));
    }

    #[test]
    fn review_title_is_bounded_and_unicode_safe() {
        let title =
            review_session_title(Some("A very long résumé parser implementation title"), 24);
        assert!(title.chars().count() <= 24);
        assert!(title.starts_with("Review: ") && title.ends_with('…'));
    }
}
